core/server/worker/src/main/java/alluxio/worker/block/TieredBlockStore.java

```diff
@@ -348,37 +494,61 @@ public void removeBlock(long sessionId, long blockId)
   public void removeBlock(long sessionId, long blockId, BlockStoreLocation location)
       throws InvalidWorkerStateException, BlockDoesNotExistException, IOException {
     LOG.debug("removeBlock: sessionId={}, blockId={}, location={}", sessionId, blockId, location);
-    removeBlockInternal(sessionId, blockId, location);
+    long lockId = mLockManager.lockBlock(sessionId, blockId, BlockLockType.WRITE);
+
+    BlockMeta blockMeta;
+    try (LockResource r = new LockResource(mMetadataReadLock)) {
+      if (mMetaManager.hasTempBlockMeta(blockId)) {
+        throw new InvalidWorkerStateException(ExceptionMessage.REMOVE_UNCOMMITTED_BLOCK, blockId);
+      }
+
+      blockMeta = mMetaManager.getBlockMeta(blockId);
+
+      if (!blockMeta.getBlockLocation().belongsTo(location)) {
+        throw new BlockDoesNotExistException(ExceptionMessage.BLOCK_NOT_FOUND_AT_LOCATION, blockId,
+            location);
+      }
+    } catch (Exception e) {
+      mLockManager.unlockBlock(lockId);
+      throw e;
+    }
+
+    try (LockResource r = new LockResource(mMetadataWriteLock)) {
+      removeBlockInternal(blockMeta);
+    } finally {
+      mLockManager.unlockBlock(lockId);
+    }
+
     synchronized (mBlockStoreEventListeners) {
       for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
         listener.onRemoveBlockByClient(sessionId, blockId);
+        listener.onRemoveBlock(sessionId, blockId, blockMeta.getBlockLocation());
       }
     }
   }
 
   @Override
   public void accessBlock(long sessionId, long blockId) throws BlockDoesNotExistException {
     LOG.debug("accessBlock: sessionId={}, blockId={}", sessionId, blockId);
-    boolean hasBlock;
     try (LockResource r = new LockResource(mMetadataReadLock)) {
-      hasBlock = mMetaManager.hasBlockMeta(blockId);
-    }
-    if (!hasBlock) {
-      throw new BlockDoesNotExistException(ExceptionMessage.NO_BLOCK_ID_FOUND, blockId);
-    }
-    synchronized (mBlockStoreEventListeners) {
-      for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
-        listener.onAccessBlock(sessionId, blockId);
+      BlockMeta blockMeta = mMetaManager.getBlockMeta(blockId);
+
+      synchronized (mBlockStoreEventListeners) {
+        for (BlockStoreEventListener listener : mBlockStoreEventListeners) {
+          listener.onAccessBlock(sessionId, blockId);
+          listener.onAccessBlock(sessionId, blockId, blockMeta.getBlockLocation());
+        }
       }
     }
   }
 
   @Override
-  public void freeSpace(long sessionId, long availableBytes, BlockStoreLocation location)
+  public synchronized void freeSpace(long sessionId, long minAvailableBytes, long maxAvailableBytes,
```

Comment:
why is this now `synchronized`?

Reply:
It now only removes blocks and shouldn't run concurrently as per allocation guarantees.